### DISCUSSION TITLE
Transaction proof gossip

### DISCUSSION THREAD

**Discussion by adiabat** [utreexo/utreexo#77] [Ideas]
There are 3 general ways to deal with proofs:

A (spam) - Just send all the proofs every time 
B (provide) - Don't send proofs to start; the node that wants proofs asks for them
C (anticipate) - Figure out what proofs your peer wants and send only what they need

Method A is the simplest certainly, but especially for post-IBD mempool, it really amplifies traffic a lot.  Method B might seem to be not too bad complexity wise but have more messages / round trip delay.  Method C seems cool but probably complicated and... ultimately it seems that you'll need to also have all the code for B inside of C anyway, since you can't perfectly predict your peer's cache.

When looking at the way bitcoin invs work though, it seems like B is the best fit, and doesn't need to add any latency from how bitcoin p2p gossip works today.

Basic idea of how it works now:

Node has tx1.  NodeA makes an inv message, with 1 inv element: type tx, tx1's txid.  Sends that inv message out to everyone, and NodeB replies with a getdata, also with type tx, tx1's txid.  Then NodeA sends NodeB a tx message, with tx1.  So you've got 3 messages: 
```
 >-- Inv -->
 <-- GetData --<
 >-- Tx -->
```
One way to implement B is to have more messages.  Send inv and getdata as normal, but the tx is a utreexo tx, which specifies the leaf positions of each of its inputs.  Then the node receiving the tx requests the proofs it needs
```
 >-- Inv -->
 <-- GetData --<
 >-- Tx -->
 <-- GetProof --<
 >-- Proof -->
```
This... is not great.  I'm sure it's doable, but here's some problems I see:
It's adding more latency.  Which is not great, but actually this is mempool tx propagation which is not super latency critical so not a huge deal.
Two new message types?  Maybe getproof is just a kind of getData.  Not a huge deal.
It changes the whole flow.  To me that's the biggest problem.  What if you send a getproof and they never send the proof?  Do you keep the tx?  Drop it?  Send getProof to other nodes?  How long do you wait?  Just seems messy to have the tx in this staging area.  It's definitely possible but I think there's a better way.

How about keeping the same flow and messages, and just modifying the messages.  Then hopefully most code can stay the same.

The relevant structs in btcd:

```
type MsgInv struct {
	InvList []*InvVect
}

type InvVect struct {
	Type InvType        // Type of data
	Hash chainhash.Hash // Hash of the data
}

type InvType uint32

const (
	InvTypeError                InvType = 0
	InvTypeTx                   InvType = 1
	InvTypeBlock                InvType = 2
	InvTypeFilteredBlock        InvType = 3
	InvTypeWitnessBlock         InvType = InvTypeBlock | InvWitnessFlag
	InvTypeWitnessTx            InvType = InvTypeTx | InvWitnessFlag
	InvTypeFilteredWitnessBlock InvType = InvTypeFilteredBlock | InvWitnessFlag
)
```

In other words, an inv message is a slice of 33 byte elements.  1 byte for a type, and 32 bytes for a hash.  It actually might be nice to change that but we can hack it if we need to.  The idea would still be like this

```
 >-- Inv -->
 <-- GetData --<
 >-- Tx -->
```

In the inv message though, each Tx would also indicate the leaf positions in uses.  The getData message requests both the tx by txid, and the proof positions the node wants (both leaf and non-leaf).  When the tx is sent, the requested proof data is sent along with it.

The nice part of this method is that most of the highest level state machine stuff stays the same.  You can think of the proofs like an optional witness, which the node can request or not.  A bridge node will never request proof.  A stump will calculate the branches up to roots and request the whole thing.  A pollard based node with some caching will request only what it needs.  There's no extra state for serving nodes to keep track of; when you send out an inv, you just stick on the leaf positions to each txid, which you know and have available already.

But how to do it exactly?  I'm open to ideas there.  One big questions: do you change the top level MsgInv and GetData definitions?  (MsgInv and MsgGetData are identical)  If you change em, you need new message types, which have variable length items, not just 33 bytes each.  If you don't change em, you have to pack things into 32 byte fields that are probably not 32 bytes, and are ints, not hashes!  

(Though really to me the like core fundamental insight in bitcoin is treating hashes as ints when you hash a block header and then do <= to it).

Keeping Inv and GetData could work like this.  Define a single new inv type, and a new flag

```
InvUtreexoFlag = 1 << 29

InvTypeProofHash    InvType = 5
InvTypeUtreexoTx    InvType = InvTypeTx | InvUtreexoFlag
```

Now when you send an inv, say with 2 txids
```
[InvTypeUtreexoTx][TXID1]
[InvTypeProofHash][packed positions1.1] (4 positions)
[InvTypeProofHash][packed positions1.2] (1 position)
[InvTypeUtreexoTx][TXID2]
[InvTypeProofHash][packed positions2.1] (3 positions)
```
Packed postions is just 4 big endian uint64s concatenated. (or you could do a cannonical / deterministic varint I guess.)  Maybe keep rules like, the positions have to be in order.  Have some big number to mean invalid, like 1<<63 or something.
The order matters; InvTypeProofHash items refer to the proofs required for the tx immediately before them.  That is not space efficient, because you might have only 1 position in positions1.2, and 3 in positions 2.1, and if you combined them you could save a whole 33 bytes.  True!  But the receiver really needs to be able to track which positions are for which txs.  Because there could be low memory nodes which grab a tx, verify it & proofs, and then keep the txid status as verified and proven, but drop all the proof data and the tx itself.  Then when they receive a block they can just look at the txids and check off all they've already seen.  Those nodes would be in trouble if they weren't sure which leaves were needed by which txids when making a getData request.

Then when the GetData happens, it's pretty much the same, but maybe ordering is not as important.  You don't need to tell the node you're requesting txs and proof hashes from which are mapped to which.
```
GetData:

[InvTypeUtreexoTx][TXID1]
[InvTypeUtreexoTx][TXID2]
[InvTypeProofHash][packed positions1] (4 positions)
[InvTypeProofHash][packed positions2] (4 positions)
```
In this case we can save that space by packing the requested positions together.  We're just requesting 2 txs and 8 hashes.

... but.. maybe we should still keep the ordering, like in the inv?  Because when you actually send the MsgTx, do you send a separate MsgProof?  Or do you stick the proof data onto the end of the MsgTx?  

If you have a separate MsgProof, then what if it doesn't show up?  Kind of annoying waiting for a separate message, it's nice if nothing's in limbo.
If you stick the proofs onto the MsgTx at the end, (MsgUtreexoTx) then you still want to keep the mapping of which hashes go to which txs.  And you probably still want the sequence.  MsgTxs coming out of order is probably already a problem in bitcoin, for cpfp txs, so maybe it's not a huge deal.  But the idea that it's kindof like segwit; we're just sticking some extra data on to the end of the transaction that's not covered in the txid... that's nice because we've done it before.  So I think a MsgUtreexoTx with proofs hashes on the end is best.

Those are my thoughts on the best way to do p2p proof gossip.  Keep inv -> getdata -> msgtx, have a new inv type with ints packed into 32 bytes instead of actual hashes.  Kind of ugly but I think it's easier than other ways to do it.  But would be great to hear other people's thoughts.

**Comment by Davidson-Souza** [comment]
> A (spam) - Just send all the proofs every time  

This is definitely not a good idea because `invs` are sent automatically, and all your peers will send the same proofs over and over again, a massive bandwidth waste, especially in bridge nodes.

>  C (anticipate) - Figure out what proofs your peer wants and send only what they need  

I believe that with set reconciliation you might be able to, but it's much more complex and evolved to code. Plus, the savings for end-users shouldn't be that high because they will end up asking for the majority of the proofs.  

>  B (provide) - Don't send proofs to start; the node that wants proofs asks for them

I think the response to getData can be changed to a new message format, as long as you keep it backwards compatible. I do like the approach taken with blocks…  
You ask for a block by sending a `getdata` with `inv` entry type `WITNESS_BLOCK | UTREEXO`. The response is a `block` message, but with a different serialization, you have the block data + witness (if any) + a [UData](https://github.com/utreexo/utreexod/blob/53dc2e1096418f7286cb29e7d0b679c3129a524a/wire/udata.go#L18-L27) part. Perhaps if someone asks for a `tx`, with `inv` type `WITNESS_TX | UTREEXO`, we could just attach proofs to the end of the buffer, just as it works with blocks. Peers may decide on whether they want proofs by tweaking the `inv` type.  
As an optimization, `getdata` might only return the positions you mentioned, but without needing to hack the `inv_message` to pack integers. I have a little concern with that though, we should be careful when implementing a node with this optimization because it might enable fingerprinting attacks against a Pollard Node, by mapping UTXOs it has cached or not. I'm not sure to what extent it might be a problem, but I think it's important to think about.  

A fourth approach would involve the `getproof` message, but you give a bunch of txs and you get a batch proof. On average, we have some overlapping and hence savings if compared to returning one proof at the time. However, it is not clear if it's worth the effort.
